### ISSUE TITLE
sharenfs documentation

### ISSUE THREAD

**Issue by clhedrick** [openzfs/zfs#11159]
### System information

Type | Version/Name
 --- |master
Distribution Name	| ubuntu
Distribution Version	| 20.3
Linux Kernel	| 5.4.0-42-generic
Architecture	|Intel 64
ZFS Version	| 0.8.3-1ubuntu12.1
SPL Version	| modinfo spl | grep -iw version

### Describe the problem you're observing

Despite many reports, documentation on sharenfs still doesn't describe what it does.

### Describe how to reproduce the problem

man zfs

### Include any warning/errors/backtraces from the system logs

Here's what I suggest: The sharenfs property can be "on", "off" or a list of options. "on" is equivalent to "rw,crossmnt".  Options other than ro, rw, and sec are accumulated into a single list, which is used for all exports. That means there is no way to use different options with different hosts or groups of hosts, other than "ro", "rw", and "sec." Most options are used as is with the Linux mount command. "ro" or "ro=hostlist" generates a readonly mount. Without a hostspec (i.e. "ro" alone) exports to all hosts. Similarly, "rw" or "rw=hostlist" generates a readwrite mount. The mounts include all options that appear in the property, except "sec", "rw" and "ro" plus the most recent version of sec appearing before the "ro" or "rw". If no "sec" option has appeared, "sec=sys" is used. hostlist may be a single hostspec or a list separated by colons. Each hostspec can be as in /etc/exports, i.e. hostname(s), IP addresses or ranges, netgroups. A leading @ in the hostspec is skipped, for compatibility with Solaris IP ranges. Thus a netgroup must be specified as @@netgroup. It does not appear to be possible to specify an IPV6 address or range, because of the use of the colon as a separator.

**Comment by clhedrick** [comment]
Incidentally, I would suggest that the inability to specify an IPV6 range is a problem, although given the use of colon, I'm not sure how to fix it. Perhaps use two colons to quote it? I also wonder whether @@netgroup was really intended. It appears that in Solaris the @ isn't needed for a netgroup (at least it's not mentioned in the documentation).

**Comment by clhedrick** [comment]
Note that the problem with unsharing from an IPV6 range is not due to ZFS. You can't manually unshare it either with "sharefs -u".

**Comment by gdevenyi** [comment]
sharenfs is pretty broken, its probably worth storing your own custom config and parsing it into an /etc/exports file.

**Comment by clhedrick** [comment]
It works just fine as long as you know how it works. We get into these discussions every time, and they go down a rathole. The feature exists. It's useful. The man page should document what it actually does. We shouldn't need ideological discussions to fix it.

I have at least one situation where this is the easiest way to do things, and others where we may need to move a pool to a different host and want the shares to move too.

**Comment by nscfreny** [comment]
I can imagine many would want to be able to set "no_root_squash" for some clients and "root_squash" for others, like the example in #9066:
```zfs set sharenfs="rw=@host1,root_squash,rw=@host2,no_root_squash" pool1/d1```

If this _is_ possible on Linux, maybe someone can point me in the right direction?
(tried with 0.7.13 and 2.0.1, has always worked on Solaris, different syntax though, not sure about FreeBSD?)

**Comment by clhedrick** [comment]
no. they scan the whole sharenfs string, saving most options globally and the sec= option for each host. root_squash is stored globally. I can see situations where you might want to be host-specific, though I don't think we have one or are likely to. It's a pretty simple change to make additional options to be treated like sec=.

**Comment by MarcusEbert** [comment]
Could such change for root_squash, or in general for other options, please be added when it is just a simple change?
While root_squash in general is fine, we need no_root_squash on a NIS server where root creates users incl. home directories on the NFS shares, as an example.
Any chance that will happen in a next version?

**Comment by gdevenyi** [comment]
We've now also run into the issue of modern unsupported options for NFS such as "pnfs" for v4.2 parallel nfs.
```
zfs set sharenfs='ro=@hostpattern*,async,pnfs' data
cannot set property for 'data': 'sharenfs' cannot be set to invalid options
```

**Comment by clhedrick** [comment]
We ended up not using sharenfs, but /etc/exports.d. I like sharenfs conceptually, but we found that service was interrupted for a substantial amount of time when making a change, while "exportfs -r" is nearly immediate. I still think it should be documented correctly.
________________________________

**Comment by ykuksenko** [comment]
The `security_label` option also seems to be unsupported.
I see: `'sharenfs' cannot be set to invalid options` on zfs `2.1.7`

**Comment by dezza** [comment]
@ykuksenko #3860 mentioned this.

They follow the options from Solaris. [Here](https://docs.oracle.com/cd/E36784_01/html/E36825/gntjt.html#SVNFSgntid) is a more recent page than the one he linked.

**Comment by afborchert** [comment]
> They follow the options from Solaris. [Here](https://docs.oracle.com/cd/E36784_01/html/E36825/gntjt.html#SVNFSgntid) is a more recent page than the one he linked.

The referenced document points to Solaris 11.2 and to the share command which is independent from ZFS and its sharing properties. This is entirely misleading. If there is any similarity between Solaris ZFS and OpenZFS than you need to refer to Solaris 10. The actual syntax and the names of ZFS properties has significantly changed in Solaris 11.  And then OpenZFS diverts also significantly from Solaris 10. In particular, the root option is not supported. And you just need one '@' for netgroups in Solaris, not two. The current OpenZFS documentation points in zfsprops(7) to exportfs(8) which in turn refers to exports(5) where you could use constructs like “@trusted(rw,no_root_squash)” which would be equivalent to Solaris 10 “root=@trusted”. But neither is supported by OpenZFS.
